File: ecos-records/src/main/java/ru/citeck/ecos/records2/utils/ValWithIdx.java
package ru.citeck.ecos.records2.utils;

import lombok.Data;
import lombok.RequiredArgsConstructor;

import java.util.function.Function;

@Data
@RequiredArgsConstructor
public class ValWithIdx<T> {

    private final T value;
    private final int idx;

    public <V> ValWithIdx<V> map(Function<T, V> mapper) {
        return new ValWithIdx<>(mapper.apply(value), idx);
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records2/utils/ValWithIdx.kt
package ru.citeck.ecos.records2.utils

class ValWithIdx<T>(val value: T, val idx: Int = 0) {

    fun <K> withValue(mapper: (T) -> K): ValWithIdx<K> {
        return ValWithIdx(mapper.invoke(value), idx)
    }
}
